Improve mto proposal description (#91)
giantswarm/rfc: multi-tenancy-observability/README.md

## Introduction

> Moving towards our observability platform, we have to consider to important topics.

Reviewer: ```suggestion
Moving towards our observability platform, we have to consider two important topics.
```

First, we generate a lot of observability data.

> For instance, we reached our ingestion limits in Prometheus more than a year ago and each new component we add brings a lot more metrics so we have to be mindful about the metrics we keep.

Reviewer: ```suggestion
For instance, we already reached our ingestion limits in Prometheus in the past and each new component we add brings a lot more metrics so we have to be mindful about the metrics we keep.
```

This issue got a whole lot bigger when we added logging with Loki to the mix.

> Second, our product presentlyallows anyone with access to Grafana on the management clusters to access all the data (metrics and logs) for all workload clusters without any kind of data seggregration. This is part of the reason why only a subset of our customers (generaly only the platform teams) have access to our managed grafana and is also why our shared installation grafana is not accessible to customers.

Reviewer: ```suggestion
Second, our product presently allows anyone with access to Grafana on the management clusters to access all the data (metrics and logs) for all workload clusters without any kind of data isolation or federation. This is part of the reason why only a subset of our customers (generaly only the platform teams) have access to our managed grafana and is also why our shared installation grafana is not accessible to customers.
```

> As part of the developer platform, we want to be able to ingest all sorts of customer observability data. This will require Giant Swarm observability platform to be able to provide data isolation between tenants.

Reviewer: ```suggestion
As part of the developer platform, we want to be able to ingest all sorts of customer observability data. For example a lot of customers are interested in monitoring of application developers workloads. This will require Giant Swarm observability platform to be able to provide data isolation between tenants, so that customer teams have limited access to the data originating on clusters they own.
```

To that end, we now need to move towards multi-tenancy in our observability.

## Current situation

### Our Logging stack

> Our logging solution Loki supports multi-tenancy based on a http header (`X-Org-ID`) and we already had to implement some basic form of multi-tenancy to be able to support the logs coming from our clusters.

Author: So right now the header value is set by the proxy. If we read data from grafana, the header is set to all tenants, but when we write, it is set based a user/password pair coming from the logging agents

> Our logging solution Loki supports multi-tenancy based on a http header (`X-Org-ID`) and we already had to implement some basic form of multi-tenancy to be able to support the logs coming from our clusters.

Author: But it will eventually move be based an the user's role in the AD :)

The graph below shows our current implementation of the multi-tenancy on logs:

> <img src="./assets/scope-orgid.png" width="500" alt="Loki multi-tenancy and X-Scope-OrgID header">

Author: What do you think about this?

```suggestion
<img src="./assets/scope-orgid.png" width="500" alt="Loki multi-tenancy and X-Scope-OrgID header">

In summary, we use a Reverse Proxy in front of all Loki components. This Proxy sets the `X-Org-ID` header based on the authentication information it receives. The proxy configuration (tenant and user/password keypairs) are managed by an operator named `logging-operator`. This operator also configure the tenant and user/password keypairs information for the logging agents that they then use to send logs to loki via the Reverse Proxy.
```

### Future of monitoring

We already have plans to migrate our monitoring stack to Grafana Mimir, which supports the same multi-tenancy mechanism as Loki. You can see the current status in this [epic](https://github.com/giantswarm/roadmap/issues/3039).

## What we want to achieve

> We want to be able to segregate access to the observability data based on the user groups they have configured in their active directory.

Reviewer: ```suggestion
We want to be able to segregate access to the observability data based on the user groups our customers have configured in their active directory.
```

<img src="./assets/current-multi-tenancy.png" width="300" alt="Loki multi-tenancy on the read path">

The read path queries `loki` providing an `X-Scope-OrgID` header which contains the identifier (the `read` tenant) of all clusters: the management cluster plus all workload clusters.

### Configuring multi-tenancy on the read path

> We have asserted that we can retrieve users groups through the OAuth token that is received from Grafana when we enable Oauth forwarding on the Loki datasource.

Reviewer: ```suggestion
We have asserted that we can retrieve users groups through the OAuth token that is received from Grafana when we enable OAuth-forwarding on the Loki datasource.
```

}
```

The component `loki-multi-tenant-proxy` which has now been renamed to `grafana-multi-tenant-proxy` should be configured with the mapping between groups of people and tenants

> Below is a proposal of such a configuration:

Reviewer: ```suggestion
Below is a proposal of the configuration of the mapping between tenants and their assigned groups from the active directory:
```

```yaml
groups:
- name: giantswarm-admins
  tenants:
  - mc-name
  - wcdev-name
  - wcprod-name
- name: Developers
  tenants:
  - wcdev-name
- name: Ops
  tenants:
  - wcdev-name
  - wcprod-name
...
```

The mapping configuration should be dynamically created and updated regardless of the data source by an operator.
The operator would be able to reconcile any kind of data source and generate the configuration expected by `grafana-multi-tenant-proxy`.

Below is a graph exposing multi-tenancy proposal:

<img src="./assets/future-multi-tenancy.png" width="300" alt="Multi-tenancy proposal on the read path">

#### Operator management

`logging-operator` is an existing component responsible for creating a secret for the multi-tenant-proxy. That secret contains a mapping between credentials and tenant (it's currently equivalent to the cluster id).

We will have to either create a new operator to handle multi-tenancy for all observability tools (`observability-multi-tenancy-operator`) or add this feature into the `logging-operator` and rename it to `observability-operator`.

We are open to suggestions on this topic.

> ### Opened questions

Reviewer: ```suggestion
### Open questions
```

}
```

The component `loki-multi-tenant-proxy` which has now been renamed to `grafana-multi-tenant-proxy` should be configured with the mapping between groups of people and tenants

Below is a proposal of the configuration of the mapping between tenants and their assigned groups from the active directory:

```yaml
groups:
- name: giantswarm-admins
  tenants:
  - mc-name
  - wcdev-name
  - wcprod-name
- name: Developers
  tenants:
  - wcdev-name
- name: Ops
  tenants:
  - wcdev-name
  - wcprod-name
...
```

The mapping configuration should be dynamically created and updated regardless of the data source by an operator.
The operator would be able to reconcile any kind of data source and generate the configuration expected by `grafana-multi-tenant-proxy`.

Below is a graph exposing multi-tenancy proposal:

<img src="./assets/future-multi-tenancy.png" width="300" alt="Multi-tenancy proposal on the read path">

#### Operator management

`logging-operator` is an existing component responsible for creating a secret for the multi-tenant-proxy. That secret contains a mapping between credentials and tenant (it's currently equivalent to the cluster id).

> We will have to either create a new operator to handle multi-tenancy for all observability tools (`observability-multi-tenancy-operator`) or add this feature into the `logging-operator` and rename it to `observability-operator`.

Author: Sure but I'd like if we make a decision together :D

> We will have to either create a new operator to handle multi-tenancy for all observability tools (`observability-multi-tenancy-operator`) or add this feature into the `logging-operator` and rename it to `observability-operator`.

Author: My opinion on this is that we should have a new observability-operator with different controllers that each take care of their own things (1 for multi-tenancy, 1 for logging, and so on) so it would be easier to see where things are going wrong without having to go in 5 different operators. @giantswarm/team-atlas what do you think?

```yaml
groups:
- name: giantswarm-admins
  tenants:
  - mc-name
  - wcdev-name
  - wcprod-name
- name: Developers
  tenants:
  - wcdev-name
- name: Ops
  tenants:
  - wcdev-name
  - wcprod-name
...
```

The mapping configuration should be dynamically created and updated regardless of the data source by an operator.
The operator would be able to reconcile any kind of data source and generate the configuration expected by `grafana-multi-tenant-proxy`.

Below is a graph exposing multi-tenancy proposal:

<img src="./assets/future-multi-tenancy.png" width="300" alt="Multi-tenancy proposal on the read path">

#### Operator management

`logging-operator` is an existing component responsible for creating a secret for the multi-tenant-proxy. That secret contains a mapping between credentials and tenant (it's currently equivalent to the cluster id).

We will have to either create a new operator to handle multi-tenancy for all observability tools (`observability-multi-tenancy-operator`) or add this feature into the `logging-operator` and rename it to `observability-operator`.

> 

Reviewer 2: ```suggestion

Atlas' current preference is towards the second option : moving towards having a single, global `observability-operator` which would embed several different controllers, each dedicated to its own observability tool (i.e one controller for logging, another one for multi-tenancy, ...). This approach has the benefit of limiting the number of different operators - and thus related repos - which would not only ease the development process but also make troubleshooting a lot whole easier as one would have only 1 operator to check.

```

Author: Thanks :)

This is the approach we are currently using for technical reasons. We do not think this would match with customers workload as they mostly have a lot of customers sharing clusters.

##### Namespace == Tenant

This approach makes it really easy for us and customers to configure tenants and it would definitely be easy to onboard teams (e.g. adding a label on the namespace would make the agents collect logs from the containers in the namespace).

But this approach brings a few questions:
1. What tenancy should common internal apps use (e.g. prometheus-operator, api-server, cilium logs are accessed by multiple internal teams)? Do we need to ensure customers are putting those apps into a namespace accessible by all?
2. What about teams that share namespaces?

##### App == Tenant

This approach is especially flexible for customers but we do not think this approach makes sense because the mapping configuration would be impossible to maintain and we would probably face technical limits pretty fast

##### Hybrid mode

This approach would allow customers to be able to set the tenancy as they want (per namespace or per app depending on the use case).

#### Implementation details

> This part is a bit gray because we are still ensure what tool we want to use. We would prefer to use only 1 as opposed to the 2 we have today (namely grafana-agent and promtail) but they are currenly not feature complete.

Reviewer: ```suggestion
The implementation details are at this stage are still an open topic, because no single tool fulfills all of our requirements. Until this RFC we use two tools, grafana-agent and promtail, which are bot not yet feature complete.
```

> We would love to use [PodLogs](https://grafana.com/docs/agent/latest/operator/api/#podlogs-a-namemonitoringgrafanacomv1alpha1podlogsa) the equivalent of a Service or PodMonitor for metrics but the existing implementation at the time of this RFC is fully compliant with our needs.

Reviewer: ```suggestion
One preferred alternative would be [PodLogs](https://grafana.com/docs/agent/latest/operator/api/#podlogs-a-namemonitoringgrafanacomv1alpha1podlogsa), the equivalent of a Service or PodMonitor for metrics, but the existing implementation at the time of this RFC is not yet fully compliant with our needs as well.
```

Existing logging agents support setting the tenant to send to loki using a custom log label.
We can configure our logging agents to get the logs of specific namespaces/pods if they have a label like `giantswarm.io/tenant: my-tenant` or using a PodLog CR (if applicable).

Pros:
- Quite easy to configure for logs

Cons:
- Tenant configuration will need to be configured in multiple places (MC for the read path and WC for the write path)
- Label approach would most likely not work for metrics OOTB (this would require customer to add new labels like metrics port, path and so on).

#### Approach 3: Configure promtail based on the MC tenancy CR

Altough the customer UX for the read path is still unclear, we can envision a CR like a Tenant CR that would map user's roles to namespaces/deployment under a given tenant name. The logging operator would then take care of applying all that logic

Pros:
- Everything is happening on the MC and it could map to the Project CR concept that has been talked in Product for a while.

Cons:
- MC needs to be aware of the state of all WCs (has all namespaces and so on). How can customers be sure they have all logs?

> #### Approach 4: We enforce 1 namespace == 1 tenant with possibility of opting out

Reviewer 3: ```suggestion
#### Approach 4: We enforce 1 namespace == 1 tenant with flexibility by app and possibility of opting out
```

It is quite easy to configure 1 tenant per namespace in our current setup and we could use a giantswarm.io/logging: false on specific namespace or containers to not get the logs.

Pros:
- Customers would only have to fill out the read path configurations (unless they want to opt-out a lot)
- With this approach, we could still have a `giantswarm.io/tenant: my-tenant` label on some apps that should have a specific tenant attached to them.

Cons:
- Our existing customers want to start slow (team by team) and opting-out for all namespaces from the start will be a hassle. An idea to avoid this would be to be able to opt-out of the multi-tenancy features with a cluster label (`giantswarm.io/tenancy: false`).

> ### Opened questions

Reviewer: ```suggestion
### Open questions
```

> - Our existing customers want to start slow (team by team) and opting-out for all namespaces from the start will be a hassle. An idea to avoid this would be to be able to opt-out of the multi-tenancy features with a cluster label (`giantswarm.io/tenancy: false`).

Author: How is that a con?

> - Our existing customers want to start slow (team by team) and opting-out for all namespaces from the start will be a hassle. An idea to avoid this would be to be able to opt-out of the multi-tenancy features with a cluster label (`giantswarm.io/tenancy: false`).

Reviewer 3: For isntance, we can't have a different tenant for network team (cilium) and observability team (prometheus-operator)

> - Our existing customers want to start slow (team by team) and opting-out for all namespaces from the start will be a hassle. An idea to avoid this would be to be able to opt-out of the multi-tenancy features with a cluster label (`giantswarm.io/tenancy: false`).

Author: But for giant swarm, I'm not sure we want to split the logs per team

> - Our existing customers want to start slow (team by team) and opting-out for all namespaces from the start will be a hassle. An idea to avoid this would be to be able to opt-out of the multi-tenancy features with a cluster label (`giantswarm.io/tenancy: false`).

Reviewer 3: Oh the approach 4 is a mix between 1 namespace = 1 tenant and 1 app = 1 tenant.
That was not clear for me because of the title

> - Our existing customers want to start slow (team by team) and opting-out for all namespaces from the start will be a hassle. An idea to avoid this would be to be able to opt-out of the multi-tenancy features with a cluster label (`giantswarm.io/tenancy: false`).

Author: Can you open a suggestion?

> - Our existing customers want to start slow (team by team) and opting-out for all namespaces from the start will be a hassle. An idea to avoid this would be to be able to opt-out of the multi-tenancy features with a cluster label (`giantswarm.io/tenancy: false`).

Author: Thanks
